eth-connector/src/lib.rs

```diff
@@ -636,38 +651,32 @@ use crate::migration::{CheckResult, InputData, Migration};
 #[cfg(feature = "migration")]
 #[near_bindgen]
 impl Migration for EthConnectorContract {
-    /// Migrate contract data
+    /// Migrate accounts balances
     #[private]
-    fn migrate(&mut self, #[serializer(borsh)] data: InputData) {
-        // Insert account
-        for (account, amount) in &data.accounts {
-            self.ft.accounts.insert(account, amount);
-        }
-        log!("Inserted accounts_eth: {:?}", data.accounts.len());
-
-        // Insert total_eth_supply_on_near
-        if let Some(total_eth_supply_on_near) = data.total_supply {
-            self.ft.total_supply = total_eth_supply_on_near;
-            log!(
-                "Inserted total_eth_supply_on_near: {:?}",
-                total_eth_supply_on_near
-            );
-        }
-
-        // Insert account_storage_usage
-        if let Some(account_storage_usage) = data.account_storage_usage {
-            self.ft.account_storage_usage = account_storage_usage;
-            log!(
-                "Inserted account_storage_usage: {:?}",
-                account_storage_usage
-            );
-        }
+    fn migrate(&mut self, #[serializer(borsh)] accounts: Vec<AccountId>) -> Promise {
+        const GAS_FOR_CALLS: Gas = Gas(140 * Gas::ONE_TERA.0);
+        ext_engine_connector::ext(self.connector.aurora_engine_account_id.clone())
+            .with_static_gas(GAS_FOR_CALLS)
+            .ft_balances_of(accounts)
+            .then(
+                ext_migrate::ext(env::current_account_id())
+                    .with_static_gas(GAS_FOR_CALLS)
+                    .migrate_callback(),
+            )
+    }
 
-        // Insert Proof
-        for proof_key in &data.used_proofs {
-            self.used_proofs.insert(proof_key);
+    fn migrate_callback(
+        &mut self,
+        #[callback]
+        #[serializer(borsh)]
+        balances: HashMap<AccountId, Balance>,
+    ) {
+        for (account, amount) in &balances {
+            if let Some(previous_balance) = self.ft.accounts.insert(account, amount) {
+                self.ft.total_supply -= previous_balance;
+            }
+            self.ft.total_supply += amount;
```

Comment:
I would move the line `self.ft.total_supply += amount;` to the beginning of the loop to prevent overflow.

Reply:
Logically, it is impossible to get overflow(underflow) here. If we got it, then it means that we did something wrong, and the migration process should be stopped. The contract will panic because the flag `overflow-checks` is enabled.